File: nucliadb/nucliadb/search/tests/integration/api/v1/test_search.py
# Copyright (C) 2021 Bosutech XXI S.L.
#
# nucliadb is offered under the AGPL v3.0 and as commercial software.
# For commercial licensing, contact us at info@nuclia.com.
#
# AGPL:
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as
# published by the Free Software Foundation, either version 3 of the
# License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program. If not, see <http://www.gnu.org/licenses/>.
#
import asyncio
import math
import os
from typing import Callable
from unittest import mock

import pytest
from httpx import AsyncClient
from nucliadb_protos.nodereader_pb2 import (
    DocumentSearchRequest,
    ParagraphSearchRequest,
    VectorSearchRequest,
)
from nucliadb_protos.writer_pb2 import Shards as PBShards

from nucliadb.common.cluster.manager import INDEX_NODES
from nucliadb.common.maindb.utils import get_driver
from nucliadb.ingest.tests.vectors import Q
from nucliadb.search.api.v1.router import KB_PREFIX
from nucliadb_models.resource import NucliaDBRoles
from nucliadb_utils.exceptions import LimitsExceededError
from nucliadb_utils.keys import KB_SHARDS

RUNNING_IN_GH_ACTIONS = os.environ.get("CI", "").lower() == "true"


@pytest.mark.asyncio
async def test_search_kb_not_found(search_api: Callable[..., AsyncClient]) -> None:
    async with search_api(roles=[NucliaDBRoles.READER]) as client:
        resp = await client.get(
            f"/{KB_PREFIX}/00000000000000/search?query=own+text",
        )
        assert resp.status_code == 404


@pytest.mark.asyncio
async def test_multiple_fuzzy_search_resource_all(
    search_api: Callable[..., AsyncClient], multiple_search_resource: str
) -> None:
    kbid = multiple_search_resource

    async with search_api(roles=[NucliaDBRoles.READER]) as client:
        resp = await client.get(
            f'/{KB_PREFIX}/{kbid}/search?query=own+test+"This is great"&highlight=true&page_number=0&page_size=20',
        )

        assert resp.status_code == 200, resp.content
        assert len(resp.json()["paragraphs"]["results"]) == 20

        # Expected results:
        # - 'text' should not be highlighted as we are searching by 'test' in the query
        # - 'This is great' should be highlighted because it is an exact query search
        # - 'own' should not be highlighted because it is considered as a stop-word
        assert (
            resp.json()["paragraphs"]["results"][0]["text"]
            == "My own text Ramon. <mark>This is great</mark> to be here. "
        )


@pytest.mark.flaky(reruns=5)
@pytest.mark.asyncio
async def test_multiple_search_resource_all(
    search_api: Callable[..., AsyncClient], multiple_search_resource: str
) -> None:
    kbid = multiple_search_resource

    async with search_api(roles=[NucliaDBRoles.READER]) as client:
        resp = await client.get(
            f"/{KB_PREFIX}/{kbid}/search?query=own+text&highlight=true&page_number=0&page_size=40",
        )

        assert resp.status_code == 200, resp.content
        assert len(resp.json()["paragraphs"]["results"]) == 40
        assert resp.json()["paragraphs"]["next_page"]
        assert resp.json()["fulltext"]["next_page"]

        shards = resp.json()["shards"]
        pos_30 = resp.json()["paragraphs"]["results"][30]
        pos_35 = resp.json()["paragraphs"]["results"][35]

        resp = await client.get(
            f"/{KB_PREFIX}/{kbid}/search?query=own+text&highlight=true&page_number=3&page_size=10&shards={shards[0]}",  # noqa
        )

        assert resp.status_code == 200, resp.content
        assert resp.json()["shards"][0] == shards[0]
        assert resp.json()["paragraphs"]["results"][0]["rid"] == pos_30["rid"]
        assert resp.json()["paragraphs"]["results"][5]["rid"] == pos_35["rid"]

        resp = await client.get(
            f"/{KB_PREFIX}/{kbid}/search?query=own+text&highlight=true&page_number=4&page_size=20",
        )
        assert resp.status_code == 200, resp.content
        for _ in range(10):
            if (
                len(resp.json()["paragraphs"]["results"]) < 20
                or len(resp.json()["fulltext"]["results"]) < 20
            ):
                await asyncio.sleep(1)
                resp = await client.get(
                    f"/{KB_PREFIX}/{kbid}/search?query=own+text&highlight=true&page_number=4&page_size=20",
                )
                assert resp.status_code == 200, resp.content
            else:
                break
        assert len(resp.json()["paragraphs"]["results"]) == 20
        assert resp.json()["paragraphs"]["total"] == 100
        assert len(resp.json()["fulltext"]["results"]) == 20
        assert resp.json()["fulltext"]["total"] == 100

        assert resp.json()["paragraphs"]["next_page"] is False
        assert resp.json()["fulltext"]["next_page"] is False


@pytest.mark.asyncio
async def test_search_resource_all(
    search_api: Callable[..., AsyncClient], test_search_resource: str
) -> None:
    kbid = test_search_resource

    async with search_api(roles=[NucliaDBRoles.READER]) as client:
        await asyncio.sleep(1)
        resp = await client.get(
            f"/{KB_PREFIX}/{kbid}/search?query=own+text&split=true&highlight=true&text_resource=true",
        )
        assert resp.status_code == 200
        assert resp.json()["fulltext"]["query"] == "own text"
        assert resp.json()["paragraphs"]["query"] == "own text"
        assert resp.json()["paragraphs"]["results"][0]["start_seconds"] == [0]
        assert resp.json()["paragraphs"]["results"][0]["end_seconds"] == [10]
        assert (
            resp.json()["paragraphs"]["results"][0]["text"]
            == "My own <mark>text</mark> Ramon. This is great to be here. "
        )
        assert len(resp.json()["resources"]) == 1
        assert len(resp.json()["sentences"]["results"]) == 1

    # get shards ids

    driver = get_driver()
    txn = await driver.begin()
    key = KB_SHARDS.format(kbid=kbid)
    async for key in txn.keys(key):
        value = await txn.get(key)
        assert value is not None
        shards = PBShards()
        shards.ParseFromString(value)
        for replica in shards.shards[0].replicas:
            node_obj = INDEX_NODES.get(replica.node)

            if node_obj is not None:
                shard = await node_obj.get_shard(replica.shard.id)
                assert shard.shard_id == replica.shard.id
                assert shard.fields == 3
                assert shard.paragraphs == 2
                assert shard.sentences == 3

                prequest = ParagraphSearchRequest()
                prequest.id = replica.shard.id
                prequest.body = "Ramon"
                prequest.result_per_page = 10

                drequest = DocumentSearchRequest()
                drequest.id = replica.shard.id
                drequest.body = "Ramon"
                drequest.result_per_page = 10

                vrequest = VectorSearchRequest()
                vrequest.id = replica.shard.id
                vrequest.vector.extend(Q)
                vrequest.result_per_page = 20
                vrequest.min_score = -1.0

                paragraphs = await node_obj.reader.ParagraphSearch(prequest)  # type: ignore
                documents = await node_obj.reader.DocumentSearch(drequest)  # type: ignore

                assert paragraphs.total == 1
                assert documents.total == 1

                vectors = await node_obj.reader.VectorSearch(vrequest)  # type: ignore

                # 0-19 : My own text Ramon
                # 20-45 : This is great to be here
                # 48-65 : Where is my beer?

                # Q : Where is my wine?
                results = [(x.doc_id.id, x.score) for x in vectors.documents]
                results.sort(reverse=True, key=lambda x: x[1])
                assert results[0][0].endswith("48-65")
                assert results[1][0].endswith("0-19")
                assert results[2][0].endswith("20-45")

    await txn.abort()


@pytest.mark.asyncio
async def test_search_pagination(
    search_api: Callable[..., AsyncClient], multiple_search_resource: str
) -> None:
    kbid = multiple_search_resource

    async with search_api(roles=[NucliaDBRoles.READER]) as client:
        n_results_expected = 100
        page_size = 20
        expected_requests = math.ceil(n_results_expected / page_size)

        results = []
        shards = None

        for request_n in range(expected_requests):
            url = f"/{KB_PREFIX}/{kbid}/search?query=own+text&highlight=true&page_number={request_n}&page_size={page_size}"  # noqa

            if shards is not None and len(shards):
                url += f"&shards={','.join(shards)}"

            resp = await client.get(url)

            assert resp.status_code == 200, resp.content

            response = resp.json()

            for result in response["paragraphs"]["results"]:
                results.append(result["rid"])

            shards = response["shards"]

            if not response["paragraphs"]["next_page"]:
                # Pagination ended! No more results
                assert request_n == expected_requests - 1

        # Check that we iterated over all matching resources
        unique_results = set(results)
        assert len(unique_results) == n_results_expected


@pytest.mark.asyncio()
async def test_resource_search_query_param_is_optional(search_api, knowledgebox_ingest):
    async with search_api(roles=[NucliaDBRoles.READER]) as client:
        kb = knowledgebox_ingest
        # If query is not present, should not fail
        resp = await client.get(f"/{KB_PREFIX}/{kb}/search")
        assert resp.status_code == 200

        # Less than 3 characters should not fail either
        for query in ("f", "fo"):
            resp = await client.get(f"/{KB_PREFIX}/{kb}/search?query={query}")
            assert resp.status_code == 200


@pytest.mark.asyncio()
async def test_search_with_duplicates(search_api, knowledgebox_ingest):
    async with search_api(roles=[NucliaDBRoles.READER]) as client:
        kb = knowledgebox_ingest
        resp = await client.get(f"/{KB_PREFIX}/{kb}/search?with_duplicates=True")
        assert resp.status_code == 200

        resp = await client.get(f"/{KB_PREFIX}/{kb}/search?with_duplicates=False")
        assert resp.status_code == 200


@pytest.fixture(scope="function")
def search_with_limits_exceeded_error():
    with mock.patch(
        "nucliadb.search.api.v1.search.search",
        side_effect=LimitsExceededError(402, "over the quota"),
    ):
        yield


@pytest.mark.asyncio()
async def test_search_handles_limits_exceeded_error(
    search_api, knowledgebox_ingest, search_with_limits_exceeded_error
):
    async with search_api(roles=[NucliaDBRoles.READER]) as client:
        kb = knowledgebox_ingest
        resp = await client.get(f"/{KB_PREFIX}/{kb}/search")
        assert resp.status_code == 402
        assert resp.json() == {"detail": "over the quota"}

        resp = await client.post(f"/{KB_PREFIX}/{kb}/search", json={})
        assert resp.status_code == 402
        assert resp.json() == {"detail": "over the quota"}

File: nucliadb_texts2/tests/test_search.rs
// Copyright (C) 2021 Bosutech XXI S.L.
//
// nucliadb is offered under the AGPL v3.0 and as commercial software.
// For commercial licensing, contact us at info@nuclia.com.
//
// AGPL:
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program. If not, see <http://www.gnu.org/licenses/>.
//

mod common;
use nucliadb_core::protos::{order_by::OrderField, order_by::OrderType, OrderBy};
use nucliadb_core::protos::{DocumentSearchRequest, Faceted, Filter};
use nucliadb_core::query_planner::{PreFilterRequest, ValidFieldCollector};
use nucliadb_core::texts::*;
use nucliadb_texts2::reader::TextReaderService;
use std::collections::HashSet;

#[test]
fn test_search_queries() {
    fn query(reader: &TextReaderService, query: impl Into<String>, expected: i32) {
        let query = query.into();
        let request = DocumentSearchRequest {
            id: "shard".to_string(),
            body: query.clone(),
            page_number: 0,
            result_per_page: 20,
            ..Default::default()
        };

        let response = reader.search(&request).unwrap();
        assert_eq!(response.total, expected, "Failed query: '{}'", query);

        assert_eq!(response.total, response.results.len() as i32);
        assert!(!response.next_page);
    }

    let reader = common::test_reader();

    // empty query matches all
    query(&reader, "", 2);

    // exact text
    query(&reader, "enough to test", 1);

    // quoted - exact text
    query(&reader, "\"enough to test\"", 1);

    // exact words
    query(&reader, "enough test", 1);

    // quoted - exact words
    query(&reader, "\"enough test\"", 0);

    // unclosed quote
    query(&reader, "\"enough test", 0);

    // additional (non existent) words
    query(&reader, "enough mischievous test", 0);

    // additional (non existent) symbols
    query(&reader, "enough - test", 0);

    // partial words
    // TODO: uncomment after fixing sc-5626
    // query(&reader, "shou", 1);
}

#[test]
fn test_prefilter_all_search() {
    let reader = common::test_reader();
    let request = PreFilterRequest {
        security: None,
        labels_formula: None,
        timestamp_filters: vec![],
        keywords_formula: None,
    };
    let response = reader.prefilter(&request).unwrap();
    assert!(matches!(response.valid_fields, ValidFieldCollector::All));
}

#[test]
fn test_prefilter_not_search() {
    let reader = common::test_reader();

    let context = nucliadb_core::query_language::QueryContext {
        field_labels: HashSet::from(["/l/mylabel".to_string()]),
        paragraph_labels: HashSet::with_capacity(0),
    };
    let query = "{ \"not\": { \"literal\": \"/l/mylabel\" } }";
    let expression = nucliadb_core::query_language::translate(Some(query), None, &context).unwrap();
    let request = PreFilterRequest {
        security: None,
        timestamp_filters: vec![],
        labels_formula: expression.labels_prefilter_query,
        keywords_formula: expression.keywords_prefilter_query,
    };
    println!("expression: {:?}", request.labels_formula);
    let response = reader.prefilter(&request).unwrap();
    let valid_fields = &response.valid_fields;
    let ValidFieldCollector::Some(fields) = valid_fields else {
        panic!("Response is not on the right variant {valid_fields:?}");
    };
    assert_eq!(fields.len(), 1);
}

#[test]
fn test_labels_prefilter_search() {
    let reader = common::test_reader();

    let context = nucliadb_core::query_language::QueryContext {
        field_labels: HashSet::from(["/l/mylabel".to_string()]),
        paragraph_labels: HashSet::with_capacity(0),
    };
    let query = "{ \"literal\": \"/l/mylabel\" }";
    let expression = nucliadb_core::query_language::translate(Some(query), None, &context).unwrap();
    let request = PreFilterRequest {
        security: None,
        labels_formula: expression.labels_prefilter_query,
        timestamp_filters: vec![],
        keywords_formula: expression.keywords_prefilter_query,
    };
    let response = reader.prefilter(&request).unwrap();
    let ValidFieldCollector::Some(fields) = response.valid_fields else {
        panic!("Response is not on the right variant");
    };
    assert_eq!(fields.len(), 1);
}

#[test]
fn test_keywords_prefilter_search() {
    let reader = common::test_reader();
    let context = nucliadb_core::query_language::QueryContext {
        field_labels: HashSet::from(["/l/mylabel".to_string()]),
        paragraph_labels: HashSet::with_capacity(0),
    };
    let labels = "{ \"literal\": \"/l/mylabel\" }";
    let keywords = "{ \"literal\": \"foobar\" }";
    let expression = nucliadb_core::query_language::translate(Some(labels), Some(keywords), &context).unwrap();
    let request = PreFilterRequest {
        security: None,
        labels_formula: expression.labels_prefilter_query,
        timestamp_filters: vec![],
        keywords_formula: expression.keywords_prefilter_query,
    };
    let response = reader.prefilter(&request).unwrap();
    let ValidFieldCollector::None = response.valid_fields else {
        panic!("Response is not on the right variant");
    };

    let expression = nucliadb_core::query_language::translate(Some(labels), None, &context).unwrap();
    let request = PreFilterRequest {
        security: None,
        labels_formula: expression.labels_prefilter_query,
        timestamp_filters: vec![],
        keywords_formula: expression.keywords_prefilter_query,
    };
    let response = reader.prefilter(&request).unwrap();
    let ValidFieldCollector::Some(fields) = response.valid_fields else {
        panic!("Response is not on the right variant");
    };
    assert_eq!(fields.len(), 1);
}

#[test]
fn test_filtered_search() {
    fn query(reader: &TextReaderService, query: impl Into<String>, filter: Filter, expected: i32) {
        let query = query.into();
        let request = DocumentSearchRequest {
            id: "shard".to_string(),
            body: query.clone(),
            filter: Some(filter),
            page_number: 0,
            result_per_page: 20,
            ..Default::default()
        };

        let response = reader.search(&request).unwrap();
        assert_eq!(response.total, expected, "Failed query: '{}'", query);

        assert_eq!(response.total, response.results.len() as i32);
        assert!(!response.next_page);
    }

    let reader = common::test_reader();

    query(
        &reader,
        "",
        Filter {
            paragraph_labels: vec![],
            field_labels: vec!["/l/mylabel".to_string()],
            labels_expression: "{ \"literal\": \"/l/mylabel\" }".to_string(),
            ..Default::default()
        },
        1,
    );
    query(
        &reader,
        "",
        Filter {
            paragraph_labels: vec![],
            field_labels: vec!["/e/myentity".to_string()],
            labels_expression: "{ \"literal\": \"/e/myentity\" }".to_string(),
            ..Default::default()
        },
        1,
    );
    query(
        &reader,
        "",
        Filter {
            paragraph_labels: vec![],
            field_labels: vec!["/l/fakelabel".to_string()],
            labels_expression: "{ \"literal\": \"/l/fakelabel\" }".to_string(),
            ..Default::default()
        },
        0,
    );
}

#[test]
fn test_search_by_field() {
    let reader = common::test_reader();

    let request = DocumentSearchRequest {
        id: "shard".to_string(),
        body: "".to_string(),
        page_number: 0,
        result_per_page: 20,
        fields: vec!["title".to_string()],
        ..Default::default()
    };

    let response = reader.search(&request).unwrap();
    assert_eq!(response.total, response.results.len() as i32);
    assert_eq!(response.total, 1);
    assert!(!response.next_page);
}

#[test]
fn test_faceted_search() {
    fn query(reader: &TextReaderService, query: impl Into<String>, facets: Faceted, expected: i32) {
        let query = query.into();
        let request = DocumentSearchRequest {
            id: "shard".to_string(),
            body: "".to_string(),
            page_number: 0,
            result_per_page: 20,
            faceted: Some(facets.clone()),
            ..Default::default()
        };
        let response = reader.search(&request).unwrap();
        println!("Response: {response:#?}");
        assert_eq!(response.total, expected, "Failed faceted query: '{}'. With facets: {:?}", query, facets);

        assert_eq!(response.total, response.results.len() as i32);
        assert!(!response.next_page);
    }

    let reader = common::test_reader();

    query(
        &reader,
        "",
        Faceted {
            labels: vec!["/".to_string(), "/l".to_string(), "/t".to_string()],
        },
        2,
    );
}

#[test]
fn test_quote_fixing() {
    fn query(reader: &TextReaderService, query: impl Into<String>) {
        let request = DocumentSearchRequest {
            id: "shard".to_string(),
            body: query.into(),
            page_number: 0,
            result_per_page: 20,
            ..Default::default()
        };

        let response = reader.search(&request).unwrap();
        assert_eq!(response.query, "\"enough test\"");
    }

    let reader = common::test_reader();

    query(&reader, "\"enough test\"");
    query(&reader, "enough test\"");
    query(&reader, "\"enough test");
}

#[test]
fn test_search_with_min_score() {
    fn query(reader: &TextReaderService, query: impl Into<String>, min_score: f32, expected: i32) {
        let query = query.into();
        let request = DocumentSearchRequest {
            id: "shard".to_string(),
            body: query.clone(),
            page_number: 0,
            result_per_page: 20,
            min_score,
            ..Default::default()
        };

        let response = reader.search(&request).unwrap();
        assert_eq!(response.results.len() as i32, expected, "Failed query: '{}'", query);
        assert!(!response.next_page);
    }

    let reader = common::test_reader();

    // Check that the min score is being used to filter the results
    query(&reader, "should", 0.0, 1);
    query(&reader, "should", 100.0, 0);
}

#[test]
fn test_int_order_pagination() {
    let reader = common::test_reader();

    let request = DocumentSearchRequest {
        id: "shard".to_string(),
        body: "".to_string(),
        page_number: 0,
        result_per_page: 1,
        order: Some(OrderBy {
            r#type: OrderType::Desc.into(),
            sort_by: OrderField::Created.into(),
            ..Default::default()
        }),
        min_score: f32::MIN,
        ..Default::default()
    };

    let response = reader.search(&request).unwrap();
    assert_eq!(response.results.len(), 1);
    assert!(response.next_page);
}

// TODO: order, timestamp filter, only_faceted, with_status
